lisp/forge-core.el

```diff
@@ -196,6 +196,41 @@ at that time."
   (let ((fn (intern (concat (symbol-name (eval type)) "--eieio-childp"))))
     `(and (fboundp ',fn) (,fn ,obj))))
 
+(defmacro forge--childcase (expr &rest clauses)
+  "Evaluate EXPR and compare to CLAUSES.
+Each clause looks like (TYPE BODY...).  EXPR is evaluated and the
+corresponding BODY is evaluated for the first element of CLAUSES
+where `forge--childp' is satisfied by EXPR and TYPE.
+
+The symbol `_' may be used as a catch-all TYPE."
+  (declare (indent 1)
+           (debug (form &rest (symbolp body))))
+  (let ((Sexpr (cl-gensym)) form)
+    (cl-block clause-loop
+      (pcase-dolist (`(,type . ,body) clauses)
+        (if (eq type '_)
+            ;; If the type is _, there's no point in continuing; the
+            ;; `cond' will always succeed at this point.  Might as
+            ;; well reduce the compile size.
```

Comment:
That would be a bug and it should result in an error. For now I am just stop being clever.